### PR TITLE
Added capability to shuffle when splitting a dataframe.

### PR DESCRIPTION
**Pull request by petiop**
This is the first part of our internal implementation extending dask-ml's `train_test_split` to shuffle dataframes within partitions (second commit will be in dask-ml). Thought we give it a shot at a contribution.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by petiop** [comment]
> Thanks for this, it seems reasonable. Can you add tests?

Thanks for the feedback @TomAugspurger. 

I extended the test that is closest to the changes made (the one testing `df.random_split`). I did not find any unit tests directly testing `pd_split`. Are those the one you'd like me to add?

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -1608,6 +1608,13 @@ def test_random_partitions():
     assert a2._name == a._name
     assert b2._name == b._name
 
+    a, b = d.random_split([0.5, 0.5], 42, True)
+    with pytest.raises(AssertionError):
+        np.testing.assert_array_equal(a.index, sorted(a.index))
+
+    a, b = d.random_split([0.5, 0.5], 42, False)
+    np.testing.assert_array_equal(a.index, sorted(a.index))
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_dataframe.py]: Move this assert up to line 1605? a and b look the same as up there.

File: dask/dataframe/tests/test_dataframe.py [hunk 2]
```diff
@@ -1608,6 +1608,13 @@ def test_random_partitions():
...
+
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_dataframe.py]: Can you also add a test to ensure the random state is passed through correctly?

```python
a1, b1 = d.random_split([0.5, 0.5], random_state=42, shuffle=True)
a2, b2 = d.random_split([0.5, 0.5], random_state=42, shuffle=True)

assert_eq(a1, a2)
assert_eq(b1, b2)
```

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -5428,25 +5431,31 @@ def cov_corr_agg(data, cols, min_periods=2, corr=False, scalar=False):
...
-    0  1  2
-    3  4  5
+    1  2  3
     4  5  6
+    2  3  4
     """
     p = list(p)
+    if shuffle:
+        if not isinstance(random_state, np.random.RandomState):
+            random_state = np.random.RandomState(random_state)
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Can this be removed?

```suggestion
```

It seems like random_state is always supplied. If so, I'd recommend making it a required argument.

Comment (petiop) [reply dask/dataframe/core.py]: If those lines are removed, `random_state` is a numpy array and `pandas.core.common.random_state` raises as it's expecting integer, `np.random.RandomState`, or `None`

Comment (TomAugspurger) [reply dask/dataframe/core.py]: Gotcha, thanks. I think pandas is being too strict there. I opened https://github.com/pandas-dev/pandas/issues/32503
